modyn/supervisor/internal/triggers/performance/decision_policy.py

```diff
@@ -0,0 +1,201 @@
+from abc import ABC, abstractmethod
+
+from modyn.config.schema.pipeline.trigger.performance.criterion import (
+    DynamicPerformanceThresholdCriterion,
+    StaticNumberAvoidableMisclassificationCriterion,
+    StaticPerformanceThresholdCriterion,
+)
+from modyn.const.types import ForecastingMethod, TriggerEvaluationMode
+from modyn.supervisor.internal.triggers.performance.data_density_tracker import (
+    DataDensityTracker,
+)
+from modyn.supervisor.internal.triggers.performance.performance_tracker import (
+    PerformanceTracker,
+)
+
+
+class PerformanceDecisionPolicy(ABC):
+    """Decision policy that will make the binary trigger decisions on
+    observations of a performance metric."""
+
+    @abstractmethod
+    def evaluate_decision(
+        self,
+        update_interval: int,
+        evaluation_scores: dict[str, float],
+        data_density: DataDensityTracker,
+        performance_tracker: PerformanceTracker,
+        mode: TriggerEvaluationMode,
+        method: ForecastingMethod,
+    ) -> bool:
+        """Evaluate the decision based on the given observation.
+
+        At the time of calling this the performance_tracker has already been updated with the new performance value
+        to allow for forecast based decisions.
+
+        Also, data_density has to be updated with the new data interval.
+
+        Args:
+            update_interval: The interval in which the decision is made.
+            performance: The observed performance metric.
+            mode: The mode in which the decision should be evaluated.
+            data_density: The data density tracker, updated with the new data interval.
+            performance_tracker: The performance tracker, updated with the new performance value.
+
+        Returns:
+            The final trigger decision.
+        """
+
+    def inform_trigger(self) -> None:
+        """Inform the decision policy that a trigger has been invoked."""
+
+
+class StaticPerformanceThresholdDecisionPolicy(PerformanceDecisionPolicy):
+    """Decision policy that will make the binary trigger decisions based on a
+    static threshold."""
+
+    def __init__(self, config: StaticPerformanceThresholdCriterion):
+        self.config = config
+
+    def evaluate_decision(
+        self,
+        update_interval: int,
+        evaluation_scores: dict[str, float],
+        data_density: DataDensityTracker,
+        performance_tracker: PerformanceTracker,
+        mode: TriggerEvaluationMode,
+        method: ForecastingMethod,
+    ) -> bool:
+        if mode == "hindsight":
+            return evaluation_scores[self.config.metric] < self.config.metric_threshold
+
+        return (evaluation_scores[self.config.metric] < self.config.metric_threshold) or (
+            performance_tracker.forecast_next_performance(mode) < self.config.metric_threshold
+        )
```

Comment:
Should we log the reason for the decision / the forecasted value somewhere?

Reply:
Good idea that might be useful! 

I cannot integrate this fully in this PR, though, as several of the follow-up PRs define code that also needs adjustment for this. If I run into problems when executing the policy, I'll come back to this and adjust the return type of `evaluate_decision` to `tuple[bool, str]`, outputting the decision reason.